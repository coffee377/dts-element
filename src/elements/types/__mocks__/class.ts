import {Type} from '../../type';

export class ClassType extends Type<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[ClassType ${this.parameters.class}]`;
  }

}
