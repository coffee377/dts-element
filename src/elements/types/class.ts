import {Container} from '../../collections';
import {emit_generics} from '../../helpers/emit-generics';
import {Type} from '../type';

// TODO: support for ClassDeclaration
export interface IClassTypeRequiredParameters {
  class: string;
}

export interface IClassTypeOptionalParameters {
  generics: Type[];
}

export class ClassType extends Type<IClassTypeRequiredParameters, IClassTypeOptionalParameters> {

  public get default_parameters(): IClassTypeOptionalParameters {
    return {
      generics: [],
    };
  }

  public _emit(_container: Container): string {
    const name = this.parameters.class;
    return `${name}${emit_generics(this.parameters.generics, this)}`;
  }

}
