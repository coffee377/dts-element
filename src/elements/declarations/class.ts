import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_object} from '../../helpers/emit-object';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ClassMember} from '../members/class';
import {ClassType} from '../types/class';
import {GenericType} from '../types/generic';

export interface IClassDeclarationRequiredParameters {
  name: string;
}

export interface IClassDeclarationOptionalParameters {
  generics: GenericType[];
  children: ClassMember[];
  extends: ClassType | null;
  abstract: boolean;
}

export class ClassDeclaration
    extends Declaration<IClassDeclarationRequiredParameters, IClassDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IClassDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      children: [],
      extends: null,
      abstract: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {children, generics} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, container);
    const generic = emit_generics(generics, this);
    const abstract = emit_abstract(this.parameters.abstract, this);
    const an_extends = emit_extends(this.parameters.extends, this);
    return `${abstract}class ${name}${generic}${an_extends} ${emit_object(children, this)}`;
  }

}
