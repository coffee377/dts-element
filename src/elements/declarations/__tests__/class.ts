jest.unmock('../class');

import {ClassDeclaration} from '../class';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new ClassDeclaration({name: 'Test'}).emit()).toMatchSnapshot();
  });
});
