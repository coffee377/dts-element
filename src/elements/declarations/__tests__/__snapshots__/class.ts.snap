// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[emit-jsdoc][emit-abstract]class [emit_declaration_name \\"Test\\"][emit-generics][emit-extends] [emit-object]"`;
